### PR TITLE
US040 - Login System

### PR DESCRIPTION
**Pull request by RobertCrabtreeKainos**

### CODE REVIEW

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        return false;
+    }
+
+    public String generateToken(String username) throws SQLException {
+        String token = UUID.randomUUID().toString();
+        Date expiry = DateUtils.addHours(new Date(), 24);
+
+        Connection c = databaseConnector.getConnection();
+
+        String insertStatement = "INSERT INTO Token (Email, Token, Expiry) " +
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: Do you need to store the token in the database? When accepting the token you should be able to verify it and use the expiry date on it to authenticate

File: src/main/java/org/kainos/ea/db/DatabaseConnector.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                throw new IllegalArgumentException("Properties file must exist and must contain user, password, name and host properties");
+            }
+
+            conn = DriverManager.getConnection("jdbc:mysql://" + host + "/" + name + "?useSSL=false", user, password);
+            return conn;
+
+        } catch (Exception e) {
+            System.err.println(e.getMessage());
+        } finally{
+            System.out.println("I will always run!");
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/DatabaseConnector.java]: Please remove

File: src/main/java/org/kainos/ea/db/AuthDao.java (hunk 1 above)
Comment (RobertCrabtreeKainos) [reply src/main/java/org/kainos/ea/db/AuthDao.java]: we say no?

File: src/main/java/org/kainos/ea/api/AuthService.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    public AuthService(AuthDao authDao) {
+        if(authDao == null)
+        {
+            throw new NullPointerException("AuthDao service provided is null");
+        }
+        this.authDao = authDao;
+    }
+
+    public String login(Login login) throws FailedToLoginException, FailedToGenerateTokenException {
+                if (authDao.validLogin(login)) {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/api/AuthService.java]: can you fix the formatting here, had to read this a couple of times :D

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Date;
+import java.util.UUID;
+
+public class AuthDao {
+    private DatabaseConnector databaseConnector;
+
+    public AuthDao(DatabaseConnector databaseConnector) {
+        if(databaseConnector == null)
+        {
+            throw new NullPointerException("Database connector passed in ctor is null");
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: typo in exception message

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public boolean validLogin(Login login) {
+        try (Connection c = databaseConnector.getConnection()) {
+            Statement st = c.createStatement();
+
+            ResultSet rs = st.executeQuery("SELECT password FROM `User` WHERE email = '"
+                    + login.getEmail() + "'");
+            while (rs.next()) {
+                return BCrypt.checkpw(login.getPassword(), rs.getString("password"));
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: Having the password check here breaks the single responsibility principle. This function is now getting the password from the db and validating it

File: src/test/java/org/kainos/ea/service/AuthServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+@ExtendWith(MockitoExtension.class)
+public class AuthServiceTest {
+    AuthDao authDao = Mockito.mock(AuthDao.class);
+    AuthService authService = new AuthService(authDao);
+    Login userLogin = new Login(
+            "email@email.com",
+            "thisIsABadPassword"
+    );
+    @Test
+    void ctor_shouldThrowNullPointerException_whenNullService() throws NullPointerException {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/service/AuthServiceTest.java]: ctor?

**Review by TomBillingtonUK — COMMENTED** [review]
You are also missing integration tests for this PR

**Review by LMBishop — APPROVED** [review]

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        return null;
+    }
+    public String generateToken(String email) throws SQLException, FailedToGetUserId {
+        int userId = getUserId(email);
+        if (userId == -1) {
+            throw new FailedToGetUserId();
+        }
+
+        String token = UUID.randomUUID().toString();
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: If you aren't using a JWT, I won't be able to approve it at the last stage as the requirements state that you must use a JWT

File: src/main/java/org/kainos/ea/db/AuthDao.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                return rs.getString("password");
+            }
+        }
+        catch (SQLException e)
+        {
+            throw new FailedToGetUserPassword();
+        }
+        return null;
+    }
+    public String generateToken(String email) throws SQLException, FailedToGetUserId, FailedToGenerateTokenException {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/db/AuthDao.java]: This function does nothing with SQL any more so probably doesn't belong here.

File: src/test/java/org/kainos/ea/integration/LoginIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+                .post(Entity.entity(validLoginRequest, MediaType.APPLICATION_JSON));
+        Assertions.assertEquals(200, response.getStatus());
+    }
+    @Test
+    void login_shouldReturn401_whenInValidLogin(){
+        Response response = APP.client().target("http://localhost:8080/api/login").request()
+                .post(Entity.entity(invalidLoginRequest, MediaType.APPLICATION_JSON));
+        Assertions.assertEquals(401, response.getStatus());
+    }
+}
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/LoginIntegrationTest.java:48]: Can you come up with a test that causes an Internal Server Error to be thrown?

File: src/test/java/org/kainos/ea/integration/LoginIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class LoginIntegrationTest {
+
+    static final DropwizardAppExtension<DropwizardWebServiceConfiguration> APP = new DropwizardAppExtension<>(
+            DropwizardWebServiceApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+    Login validLoginRequest = new Login("admin@admin.com", "securePassword");
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/integration/LoginIntegrationTest.java]: If this is a valid login to the database, it's not very secure to have the credentials available to the public in the source code

File: src/test/java/org/kainos/ea/auth/TokenServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import org.kainos.ea.db.AuthDao;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.sql.SQLException;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/auth/TokenServiceTest.java:20]: There are a few scenarios you aren't testing. can you cover them

File: src/test/java/org/kainos/ea/dao/AuthDaoTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        Mockito.when(connectionMock.createStatement()).thenReturn(statementMock);
+        Mockito.when(statementMock.executeQuery(any(String.class))).thenReturn(resultSetMock);
+        Mockito.when(resultSetMock.next()).thenReturn(true);
+        Mockito.when(resultSetMock.getInt(1)).thenReturn(1);
+
+        int result = authDao.getUserId("email@email.com");
+        assertEquals(result, 1);
+    }
+    @Test
+    public void generateToken_shouldThrowFailedToGetUserIdException_whenGetUserIdReturnNegative1() throws SQLException {
```
Comment (TomBillingtonUK) [inline src/test/java/org/kainos/ea/dao/AuthDaoTest.java]: This should be in the token service test file

File: src/main/java/org/kainos/ea/auth/JWTService.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package org.kainos.ea.auth;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.algorithms.Algorithm;
+import org.apache.commons.lang3.time.DateUtils;
+
+import java.util.Date;
+
+public class JWTService {
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/auth/JWTService.java:9]: I know why you did this, however this file needs tests now

File: src/main/java/org/kainos/ea/auth/TokenService.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    private AuthDao authDao;
+    private JWTService jwtTokenService;
+
+
+    public TokenService(AuthDao authDao, JWTService jwtTokenService)
+    {
+        if(authDao == null || jwtTokenService == null)
+        {
+            throw new NullPointerException("Null service passed into constructor.");
+        }
```
Comment (TomBillingtonUK) [inline src/main/java/org/kainos/ea/auth/TokenService.java]: Put a similar comment on Leo's UI. Can you, as a team, come up with a consistent way of handling these null checks? I don't mind which way you go as long as it's consistent

File: .idea/workspace.xml
Comment (shaunganley) [inline .idea/workspace.xml:1]: shouldn't this file be blocked by gitignore?

**Review by shaunganley — APPROVED** [review]

File: .idea/workspace.xml
Comment (RobertCrabtreeKainos) [reply .idea/workspace.xml:1]: When it merges to master it will be ignored.

**Pull request merged**